### PR TITLE
add TLS support for NLB / fix several NLB bugs

### LINKED ISSUE #69264: It is impossible to deploy identical NLB backed service in two different K8S clusters in same AWS region
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind feature

**What happened**:
I've tried to switch two Istio ingressgateways in different K8S clusters (in the same AWS region) to NLB loadbalancers, but receive error:

> error processing service istio-system/istio-ingressgateway (will retry): failed to ensure load balancer for service istio-system/istio-ingressgateway: Error creating load balancer target group: "DuplicateTargetGroupName: A target group with the same name 'k8s-tg-9ff6919e87f45-80-31380' exists, but with different settings

**What you expected to happen**:
Second loadbalancer's targetgroups created successfully.

**How to reproduce it (as minimally and precisely as possible)**:
Create two K8S clusters in the same AWS region. Create in each cluster service backed by NLB loadbalancer with the same namespace/servicename and ports combination.

**Anything else we need to know?**:
Problem is related to target group name calculation in function createTargetName:
https://github.com/kubernetes/kubernetes/blob/bcc23dfdc1620a2bbe840d62b739bcd1bda260b7/pkg/cloudprovider/providers/aws/aws_loadbalancer.go#L330-L333
Names for same combination of namespace/servicename (istio-system/istio-ingressgateway) and ports in different clusters in same region will generate identical names for target groups. 

**Environment**:
- Kubernetes version (use `kubectl version`): 1.11.2
- Cloud provider or hardware configuration: AWS
- Install tools: kops

### LINKED ISSUE #73297: Support TLS termination with AWS NLB
**What would you like to be added**:

AWS has announced TLS termination for network load balancers: https://aws.amazon.com/blogs/aws/new-tls-termination-for-network-load-balancers/

It would be amazing if Kuberentes could support this! The annotations are all defined, we would need to hook up the cloud provisioner.

```yaml
service.beta.kubernetes.io/aws-load-balancer-type: nlb
service.beta.kubernetes.io/aws-load-balancer-ssl-cert: `arn:...`
```

**Why is this needed**:

ACM can only be used with AWS Load balancers, so supporting TLS with NLBs means we will be able to leverage ACM and support http2 and WebSockets without using alb-ingress. This would resolve my need for ALBs in Kubernetes: https://github.com/kubernetes/kubernetes/issues/30518. For example it would make running nginx-ingress with http2 and WebSocket support possible with ACM.

### LINKED ISSUE #75006: AWS NLB update service with multiple ports fails.
**What happened**:
There are multiple bugs when update NLB service with multiple ports:
1. [new targetGroup created are not tagged](https://github.com/kubernetes/kubernetes/blob/feb0937fa4433df5fb3b38f7d6afa98d17b0d486/pkg/cloudprovider/providers/aws/aws_loadbalancer.go#L231)
1. multiple new targetGroup will result in error since [ELBV2 only supports tagging 1 resource per request](https://github.com/kubernetes/kubernetes/blob/feb0937fa4433df5fb3b38f7d6afa98d17b0d486/pkg/cloudprovider/providers/aws/aws_loadbalancer.go#L314)

**What you expected to happen**:
The above two bugs are fixed.

**How to reproduce it (as minimally and precisely as possible)**:
1. create an service with NLB.
1. modify the service with 1 additional port. (will observe new TG created but not tagged)
1. modify the service with 2 additional port.(will observe error message says `ELBV2 only supports tagging 1 resource per reques`] 

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by M00nF1sh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add [TLS support](https://aws.amazon.com/blogs/aws/new-tls-termination-for-network-load-balancers/) for NLB
Fix several NLB bugs(around targetGroup naming/tagging)

**Which issue(s) this PR fixes**:

Fixes #73297
Fixes #69264
Fixes #75006

**Special notes for your reviewer**:
1. new targetGroups will get name k8s-{namespace:8}-{name:8}-{uuid:10}. 
2. TLS is an upgrade version of SSL protocol, in CLB, both SSL/TLS is identified as `SSL`, however, in ALB/NLB, both SSL/TLS is identified as `TLS`. To avoid confusing and ease migration from CLB to NLB, `service.beta.kubernetes.io/aws-load-balancer-backend-protocol:ssl` is re-used for denoting backend SSL in NLB as well.
3. Test done:
    * migration from TCP to TLS termination:
        1. create NLB service with TCP port(443), which forward to backend HTTPS port(443). 
        1. access the NLB, observed TLS termination at backend works fine(cert by backend).
        1. migrate to  NLB TLS termination by adding two annotation: `service.beta.kubernetes.io/aws-load-balancer-ssl-cert: arnOfACMCert` and `service.beta.kubernetes.io/aws-load-balancer-backend-protocol: ssl`
        1. access the NLB, observed TLS termination at NLB works fine(cert by ACM), and NLB still talks to backend in TLS 😄 .
        1. migrate to TCP backend by remove annotation `service.beta.kubernetes.io/aws-load-balancer-backend-protocol: ssl`, and change service targetPort to an HTTP port.
        1. access the NLB, observed TLS termination at NLB works fine(cert by ACM), and NLB still talks to backend in TCP(HTTP) 😄 .
        1. migrate back to TLS backend by add annotation `service.beta.kubernetes.io/aws-load-balancer-backend-protocol: ssl`, and change service targetPort to an TLS port
        1. access the NLB, observed TLS termination at NLB works fine(cert by ACM), and NLB still talks to backend in TCP(HTTP) 😄 .

     * create NLB service with multiple TLS/TCP port.
        1. create NLB service with TCP port(80), which forward to backend HTTP port(80), and TLS port(443) which forward to backend HTTP port(80), by adding annotation `service.beta.kubernetes.io/aws-load-balancer-ssl-cert: arnOfACMCert`, and `service.beta.kubernetes.io/aws-load-balancer-ssl-ports: 443`
         1. observed both port(80:TCP, 443:TLS) works as expected.
     * add TLS port to existing NLB service.
        1. create NLB service with TCP port(80), which forward to backend HTTP port(80).
        1. add an TLS port(443), which forward to backend HTTP port(80) by adding annotation `service.beta.kubernetes.io/aws-load-balancer-ssl-cert: arnOfACMCert`, and `service.beta.kubernetes.io/aws-load-balancer-ssl-ports: 443`
        1. observed both port(80:TCP, 443:TLS) works as expected.
     * modify SSL policy
        1. create NLB service with TLS port.
        1. modify SSL policy by adding annotation `service.beta.kubernetes.io/aws-load-balancer-ssl-negotiation-policy: ELBSecurityPolicy-2016-08`
        1. verify SSL policy updated.
      * targetGroup names/ tagging
        1. during the above testing, all new targetGroups are tagged properly, and named properly. targetGroups are recreated as need(while old targetGroup get deleted).
    

**Does this PR introduce a user-facing change?**:

```release-note
Add TLS termination support for NLB
```

### CODE REVIEW

**Comment by usvgmani** [comment]
Do we know when this bug will be fixed? We are interested in installing Istio on our AWS EKS with multiple NLBs --> Fixes #69264

**Comment by evanfuller** [comment]
Any updates? I'm also interested in this feature.

File: pkg/cloudprovider/providers/aws/aws_loadbalancer.go [hunk 1]
```diff
@@ -210,66 +190,100 @@ func (c *Cloud) ensureLoadBalancerv2(namespacedName types.NamespacedName, loadBa
...
-					if targetGroup, ok := nodePortTargetGroup[nodePort]; !ok {
-						// Create new Target group
-						targetName := createTargetName(namespacedName, frontendPort, nodePort)
+					listenerNeedsModification := false
+
+					if aws.StringValue(listener.Protocol) != mapping.FrontendProtocol {
+						listenerNeedsModification = true
+					}
+					switch mapping.FrontendProtocol {
+					case "TLS":
```
Comment (christopherhein) [inline pkg/cloudprovider/providers/aws/aws_loadbalancer.go]: Should `"TLS"` and `"TCP"` be `consts`? also should it be `"SSL"`?

**Comment by yurrriq** [comment]
Any update or timeline for this getting merged?

**Comment by lelabo-marc** [comment]
@gnufied @jsafrane ?

**Review by micahhausler — APPROVED** [review]
/approve
/lgtm

**Comment by yurrriq** [comment]
What are the chances this'll make its way into a 1.11.x release?

**Comment by matthewejohnson711** [comment]
Dragging in on the mail chain.

**Comment by M00nF1sh** [comment]
> What are the chances this'll make its way into a 1.11.x release?

Hi, I'll check whether there are merge conflicts and cherry pick this back 😄

**Comment by micahhausler** [comment]
Sorry, cherrypicks are only for bugfixes not features

**Comment by yurrriq** [comment]
To me, this seems like a bit of both. So which version will have these changes then?

**Comment by tnachen** [comment]
+1 What version is this going into?

**Comment by M00nF1sh** [comment]
@tnachen Hi, this will go into v1.15 😄

**Comment by yurrriq** [comment]
Thanks for the response, @M00nF1sh. I guess it's time to put the heat on the kops team to catch up then :)

**Comment by davidxjohnson** [comment]
Thanks @M00nF1sh for working on this.

**Comment by igorvpcleao** [comment]
Hi @M00nF1sh and @micahhausler!
Do you know why this has not been released?

**Comment by M00nF1sh** [comment]
> Hi @M00nF1sh and @micahhausler!
> Do you know why this has not been released?

@igorvpcleao Hi, this is already available in k8s v1.15.

**Comment by yogeek** [comment]
Can someone confirm this has been released in 1.15 please ?
Because I am having the #69264 issue when trying to configure Istio NLB for 2 clusters in the same zone with the following versions :
```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:54Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.5", GitCommit:"20c265fef0741dd71a66480e35bd69f18351daea", GitTreeState:"clean", BuildDate:"2019-10-15T19:07:57Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```
(cluster has been deployed on AWS by kubeadm 1.15.3)

**Comment by chris-mac** [comment]
Is anybody using a sane workaround for EKS + NLB + ACM until v1.15 + becomes available for EKS?

**Comment by lelabo-marc** [comment]
> Is anybody using a sane workaround for EKS + NLB + ACM until v1.15 + becomes available for EKS?

Also interested ...

**Comment by danielelisi** [comment]
> Is anybody using a sane workaround for EKS + NLB + ACM until v1.15 + becomes available for EKS?

For now I'm adding the ACM certs manually via CLI/UI after the NLB is provisioned

**Comment by ksgithub** [comment]
> > Is anybody using a sane workaround for EKS + NLB + ACM until v1.15 + becomes available for EKS?
> 
> For now I'm adding the ACM certs manually via CLI/UI after the NLB is provisioned

AWS EKS now is supports 1.15 for upgrade and can use this already

**Pull request merged**